src/UnoFlipModel.java

```diff
@@ -12,68 +12,330 @@
-        currentTurn = 0; // start from player 1
-        nextPlayerIndex = 1; // next player is player 2
-        deck = new Deck(); //create a new deck
-        currentColour = null; //set colour to null
-        currentRank = null; //set rank to null
-        //userInput = new Scanner(System.in); //scanner used for user input
-        numPlayers = 0; // initialize to 0
-        chosenCardIndex = -1; //initialize to -1
-        turnSeqs = new ArrayList<TurnSequence>(); //arraylist for turn sequences
-        skipTurn = false;
-        dontAsk = false;
-        for(int i =0;i<=8;i++){
-            turnSeqs.add(new Number(this)); //Number
+    public void notifyViews(){
+
+        //make sure there are views in the view arraylist to send UnoFlipEvents to
+        if(!this.views.isEmpty()){
+
+
+            //Sends events to the view to update based on different game situations
+            // If the current top card is a Wild Draw 2 and the next player declines to challenge (dontAskChallenge flag prevents repetition of this situation)
+            if(this.topCard.isWild() && !this.status.equals(STATUS_CHALLENGE_INNOCENT) && !this.status.equals(STATUS_CHALLENGE_GUILTY) && !this.dontAskChallenge){
+
+                this.status = this.currentColour.toString(); //set status as the current colour chosen by the player (ex: RED)
+
+                for( UnoFlipView view: this.views ) {
+                    view.handleUnoFlipStatusUpdate( new UnoFlipEvent(this, getCurrentPlayer().getName(), this.topCard.toString(), getCurrentPlayer().toString(),this.status,(this.currentRank == Card.Rank.WILD || this.currentRank == Card.Rank.WILD_DRAW_2)));
+                }
+
+            } else {
+
+                for( UnoFlipView view: this.views ) {
+                    view.handleUnoFlipStatusUpdate( new UnoFlipEvent(this, getCurrentPlayer().getName(), this.topCard.toString(), getCurrentPlayer().toString(),this.status ,this.currentRank == Card.Rank.WILD ));
+                }
+            }
         }
-        turnSeqs.add(new DrawOne(this)); //Draw_One
-        turnSeqs.add(new Reverse(this)); //Reverse
-        turnSeqs.add(new Skip(this)); //Skip
-        turnSeqs.add(new Wild(this)); //Wild
-        turnSeqs.add(new WildDrawTwo(this)); //Wild Draw Two
-        turnSeqs.add(new SelfDrawOne(this)); //Self Draw One
+    }
 
 
-        this.views = new ArrayList<UnoFlipView>();
+    /**
+     * PlayTurn method is used to handle game logic request sent by UnoFlipController for when a card is placed by the player
+     * or when the player draws a card.
+     * @param btnIndex - the index of the cards that is being played, -1 if player draws a card
+     */
+    public void playTurn(int btnIndex){
+
+        //if player has not played/drawn a card yet, allow player to play a card or draw from the deck
+        if (!this.turnFinished) {
+            this.chosenCardIndex = btnIndex;
+
+            //if player clicks on the draw one button
```

Comment:
I have mentioned this a few times in our review, avoid coupling the model with the view in comments